subsys/debug/asan.c

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2019 Jan Van Winkel <jan.van_winkel@dxplore.eu>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_SOC_POSIX)
+#include <sanitizer/lsan_interface.h>
+#include "soc.h"
+#endif
+
+const char *__asan_default_options()
+{
+	return
+#ifdef CONFIG_TEST
+		"log_path=stdout:"
+#endif
+#if defined(CONFIG_64BIT) && defined(__GNUC__) && !defined(__clang__)
+		/* Running leak detection at exit could lead to a deadlock on
+		 * 64-bit boards if GCC is used.
+		 * https://github.com/zephyrproject-rtos/zephyr/issues/20122
+		 */
+		"leak_check_at_exit=0:"
+#endif
+		"";
+}
+
+#if defined(CONFIG_SOC_POSIX)
+
+static void lsan_do_leak_check(void)
+{
+	__lsan_do_leak_check();
```

Comment:
Note: I'm still not too comfy with hooking into this asan interface always.

Defining __asan_default_options would seem harmless enough even if the interface changes, but calling this one would seem to be more risky, as it not only depends on the interface not being changed, but also depends on what [this function does](https://github.com/gcc-mirror/gcc/blob/master/libsanitizer/include/sanitizer/lsan_interface.h#L43).

Anyhow, please see my question above about it being needed at all.

Reply:
I can see your concern but if this code is removed we lose the leak detection support from ASAN for native_posix_64 boards if they are build with GCC.

Reply:
Yep.. but note my previous comment about its limited value